pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -812,129 +1129,116 @@ func (ds *sqlPlugin) ListMatchingEntries(ctx context.Context,
+	}, nil
 }
 
-func (ds *sqlPlugin) DeleteToken(ctx context.Context, req *datastore.JoinToken) (*common.Empty, error) {
+func deleteToken(tx *gorm.DB, req *datastore.JoinToken) (*common.Empty, error) {
 	var t JoinToken
+	if err := tx.Find(&t, "token = ?", req.Token).Error; err != nil {
+		return nil, sqlError.Wrap(err)
+	}
 
-	err := ds.db.Find(&t, "token = ?", req.Token).Error
-	if err != nil {
-		return &common.Empty{}, err
+	if err := tx.Delete(&t).Error; err != nil {
+		return nil, sqlError.Wrap(err)
 	}
 
-	return &common.Empty{}, ds.db.Delete(&t).Error
+	return &common.Empty{}, nil
 }
 
-// PruneTokens takes a Token message, and deletes all tokens which have expired
-// before the date in the message
-func (ds *sqlPlugin) PruneTokens(ctx context.Context, req *datastore.JoinToken) (*common.Empty, error) {
-	var staleTokens []JoinToken
-	resp := new(common.Empty)
-
-	err := ds.db.Where("expiry <= ?", req.Expiry).Find(&staleTokens).Error
-	if err != nil {
-		return resp, err
-	}
-
-	for _, t := range staleTokens {
-		err := ds.db.Delete(&t).Error
-		if err != nil {
-			return resp, err
-		}
+func pruneTokens(tx *gorm.DB, req *datastore.JoinToken) (*common.Empty, error) {
+	if err := tx.Where("expiry <= ?", req.Expiry).Delete(&JoinToken{}).Error; err != nil {
+		return nil, sqlError.Wrap(err)
 	}
 
-	return resp, nil
+	return &common.Empty{}, nil
 }
 
-func (ds *sqlPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
-	resp := &spi.ConfigureResponse{}
-
-	// Parse HCL config payload into config struct
-	config := &configuration{}
-	hclTree, err := hcl.Parse(req.Configuration)
+// modelToBundle converts the given bundle model to a Protobuf bundle message. It will also
+// include any embedded CACert models.
+func modelToBundle(model *Bundle) (*datastore.Bundle, error) {
+	id, err := idutil.ParseSpiffeID(model.TrustDomain, idutil.AllowAnyTrustDomain())
 	if err != nil {
-		resp.ErrorList = []string{err.Error()}
-		return resp, err
+		return nil, sqlError.Wrap(err)
 	}
-	err = hcl.DecodeObject(&config, hclTree)
-	if err != nil {
-		resp.ErrorList = []string{err.Error()}
-		return resp, err
+
+	caCerts := []byte{}
+	for _, c := range model.CACerts {
+		caCerts = append(caCerts, c.Cert...)
 	}
 
-	if config.DatabaseType == "" {
-		return resp, errors.New("database_type must be set")
+	pb := &datastore.Bundle{
+		TrustDomain: id.String(),
+		CaCerts:     caCerts,
 	}
 
-	if config.ConnectionString == "" {
-		return resp, errors.New("connection_string must be set")
+	return pb, nil
+}
+
+func validateRegistrationEntry(entry *common.RegistrationEntry) error {
+	if entry.Selectors == nil || len(entry.Selectors) == 0 {
+		return sqlError.New("invalid registration entry: missing selector list")
 	}
 
-	if config.ConnectionString != ds.ConnectionString {
-		ds.DatabaseType = config.DatabaseType
-		ds.ConnectionString = config.ConnectionString
-		return resp, ds.restart()
+	if len(entry.SpiffeId) == 0 {
```

Comment:
should we do a more stringent validation check here? registration entries should be workload entries that belong to our trust domain

Reply:
that would certainly be defensive. however those checks belong in the core and I'm not sure it is realistic to ask plugin authors to do that kind of validation. If so, we need to write some more detailed documentation on the semantics of these functions (e.g. which functions fail on a fetch if the row does not exist... not all of them do)